TFUtil.py

```diff
@@ -2688,24 +2688,41 @@ def reversed(x):
   return y
 
 
-def flatten_with_seq_len_mask(x, seq_lens, time_major=False):
+def flatten_with_seq_len_mask(x, seq_lens, batch_dim_axis=None, time_dim_axis=None, time_major=None):
   """
-  :param tf.Tensor x: shape (batch,time,...s...) with time_major=False or otherwise shape (time,batch,...s....)
+  :param tf.Tensor x: shape (batch,...s..., time, ...s'...) or shape (time,...s...., batch, ...s'...)
   :param tf.Tensor seq_lens: shape (batch,) of int32
-  :param bool time_major: if the time-dim is the first dimension in x
-  :return: tensor of shape (time', ...s...) where time' = sum(seq_len) <= batch*time
+  :param int batch_dim_axis: index of batch_dim in x
+  :param int time_dim_axis: index of time_dim in x
+  :param bool time_major: whether time axis is 0 (redundant, kept for compatibility)
+  :return: tensor of shape (time', ...s...s'...) where time' = sum(seq_len) <= batch*time
   :rtype: tf.Tensor
   """
+  # time_major is set(old_variant) => batch_dim_axis and time_dim_axis have to be None
+  assert (batch_dim_axis is None and time_dim_axis is None) or time_major is None
+  assert (batch_dim_axis is not None and time_dim_axis is not None) or time_major is not None
+  if time_major is not None:
+    if time_major:
+      batch_dim_axis, time_dim_axis = 1, 0
+    else:
+      batch_dim_axis, time_dim_axis = 0, 1
```

Comment:
I don't understand, where to put the assert. Right after the if expression?

Reply:
Just that it is executed in all possible cases. Right after the `if` expression. Isn't that clear from the code? You are just verifying, that in all possible cases, `batch_dim_axis is not None and time_dim_axis is not None` (because you just have set it, or you expect it to be set from the input). So from now on, you don't have to care about that. That makes code simpler to read.